source/common/router/router.cc

```diff
@@ -746,6 +746,15 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
       !transport_socket_options_ || !transport_socket_options_->http11ProxyInfo().has_value();
   UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(
       *this, std::move(generic_conn_pool), can_send_early_data, can_use_http3);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() &&
+      Runtime::runtimeFeatureEnabled("envoy.restart_features.ensure_connection_retry")) {
+    upstream_request->addUpstreamCallbacks(*this);
+    wait_for_connect_ = true;
+    retry_connection_failure_buffer_limit_ =
+        std::max(retry_shadow_buffer_limit_, 2 * cluster_->perConnectionBufferLimitBytes());
```

Comment:
why 2x? Please comment inline
If you need to keep this as-is please handle and regression test overflow.

Reply:
I already added a test: see ConnectionRetryBufferOverflow:
https://github.com/DataDog/envoy/blob/dfbcf58749b47bc8f98241d78023975da0b94f34/test/integration/integration_test.cc#L2473

Reply:
why worry about reaching this limit before the connection is established?
If we're willing to buffer 5m upstream I don't think we should start out buffering 10M before connection establishment.

For overflow I meant integer overflow - I think if we keep the 2x you can actually overflow the uint32_t

Reply:
That make sense, I did not thought of the integer overflow.  I will only increase it by 1MiB and  Checking that `cluster_->perConnectionBufferLimitBytes()<std::numeric_limits<uint32_t>::max()-1048576`